Add new cargo.md tutorial and standalone test projects (#2924)
unicode-org/icu4x: docs/tutorials/cargo.md

> # Setting up your Cargo.toml file for ICU4X

Reviewer: thought: increasingly of the opinion that we should be making an MdBook that contains an organized set of tutorials as well as deeper docs on e.g. stability policies

Author: Agreed; the docs folder gives us a lot of material we can use to start it

edition = "2021"
publish = false

[dependencies]
icu = "1.0"
icu_testdata = "1.0"
```

In your main.rs, you can use most functionality of ICU4X.

[« Fully Working Example »](./cargo_tests/testdata)

## Cargo.toml with experimental testdata

If you wish to use an experimental feature, such as `icu_segmenter`, set up your Cargo.toml like this:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
icu = { version = "1.0", features = ["experimental"] }
icu_testdata = { version = "1.0", features = ["icu_segmenter"] }
```

In your main.rs, you can now use experimental features.

[« Fully Working Example »](./cargo_tests/experimental)

## Cargo.toml with Buffer Provider

If you wish to generate your own data in blob format and pass it into ICU4X, enable the "serde" Cargo feature as follows:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
icu = { version = "1.0", features = ["serde"] }
icu_provider_blob = "1.0"
```

To generate your blob data, first run `cargo build` before you run `icu4x-datagen`.

[« Fully Working Example »](./cargo_tests/buffer)

## Cargo.toml with Baked Provider

> If you wish to bake in data auto-generated in your build.rs, you need to add additional dependencies as follows:

Reviewer: issue: I'm a bit wary of recommending people call datagen (which is super heavy) in a build script without adding all the caveats that it is super heavy and linking to a dedicated tutorial about the various options here.

Reviewer: also it's not normal for build scripts to hit the network, and in particular people using non-cargo build systems _hate_ this (but it's been very abnormal in Rust even before that. People don't like nondeterminism in their build). I'm not sure we should be recommending this model, especially since most users will want to generate-once-and-check-in.

Author: CC @robertbastian 

Valid points. It _is_ super convenient though that databake-datagen is capable of being run in build.rs. There are examples of build.rs being _heavy_, so I'm not super worried about that; actually, this build.rs is fairly fast given a limited set of keys and locales. Being _non-deterministic_ is a bit trickier; at least worth a note that you probably don't want to do this unless you are a leaf crate.

Also, this model is already suggested in passing in data_management.md, and I think we've discussed it before. I'm not intending to introduce anything new here; just adding an example of something you can already do.

Reviewer: Yeah I think we should _document_ this approach somewhere but I do not think we should be endorsing it too much. It's fine for now, perhaps we can clean this up in a followup.

Reviewer 2: Well, the network only gets hit if you call `SourceData::with_{icuexport}_{tag,latest}`, which clearly document that they're downloading data. If you don't use those, datagen is offline.

# Setting up your Cargo.toml file for ICU4X

ICU4X makes heavy use of small crates and Cargo features in order to be highly modular. This tutorial is intended to help you set up a Cargo.toml file to download what you need for ICU4X.

## Basic Cargo.toml with testdata

The most basic Cargo.toml to get you off the ground is the following:

```toml
[package]
name = "demo_testdata"
version = "0.1.0"
edition = "2021"

> publish = false

Reviewer 2: I don't think these keys are very relevant to the example, given that you also link to the actual `Cargo.toml`.

Author: Do you suggest I remove the entire `[package]` section, or only the `publish =`?

[dependencies]
icu = "1.1"
icu_testdata = "1.1"
```

> In your main.rs, you can use most functionality of ICU4X.

Reviewer 2: 'most' is pretty handwavy. You can use all stable functionality, for some locales (link to test locales?).

Author: You also can't use certain features that are not re-exported, like most of the data provider crates.

[dependencies]
icu = "1.1"
icu_provider = "1.1" # for databake
litemap = "0.6" # for databake
zerovec = "0.9" # for databake

# for build.rs:
[build-dependencies]
icu = "1.1"
icu_datagen = "1.1"
icu_provider = "1.1"
```

This example has an additional section for auto-generating the data in build.rs. In your build.rs, invoke the ICU4X Datagen API with the set of keys you require. Don't worry; if using databake, you will get a compiler error if you don't specify enough keys.

Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section,  invoke `icu4x-datagen` manually, and potentially check in the output to your version control. See [data_management.md](./data_management.md) for more information.

[« Fully Working Example »](./cargo_tests/baked)

## Cargo.toml with Thread Safety (Lazy Static)

If you wish to share ICU4X objects between threads, you must enable the `"sync"` Cargo feature of the `icu_provider` crate, as follows:

```toml
[package]
name = "demo_lazy_static"
version = "0.1.0"
edition = "2021"
publish = false

> # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

Reviewer 2: remove

# Setting up your Cargo.toml file for ICU4X

ICU4X makes heavy use of small crates and Cargo features in order to be highly modular. This tutorial is intended to help you set up a Cargo.toml file to download what you need for ICU4X.

## Basic Cargo.toml with testdata

The most basic Cargo.toml to get you off the ground is the following:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
icu = "1.0"
icu_testdata = "1.0"
```

In your main.rs, you can use most functionality of ICU4X.

[« Fully Working Example »](./cargo_tests/testdata)

## Cargo.toml with experimental testdata

If you wish to use an experimental feature, such as `icu_segmenter`, set up your Cargo.toml like this:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]

> icu = { version = "1.0", features = ["experimental"] }

Reviewer 2: Make sure to update this after your other PR

Author: Also requires 1.1

Author: Done

icu_testdata = { version = "1.0", features = ["icu_segmenter"] }
```

In your main.rs, you can now use experimental features.

[« Fully Working Example »](./cargo_tests/experimental)

## Cargo.toml with Buffer Provider

If you wish to generate your own data in blob format and pass it into ICU4X, enable the "serde" Cargo feature as follows:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
icu = { version = "1.0", features = ["serde"] }
icu_provider_blob = "1.0"
```

> To generate your blob data, first run `cargo build` before you run `icu4x-datagen`.
> 
> For more information on building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).

Reviewer 2: Building before datagen is only required with the `keys-from-bin` option. Let's defer to the datagen tutorial.

```suggestion
To learn about building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).
```

[« Fully Working Example »](./cargo_tests/buffer)

## Cargo.toml with Baked Provider and build.rs

If you wish to use baked data for ICU4X, add some additional dependencies as follows:

```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
icu = "1.0"
icu_provider = "1.0" # for databake

> litemap = "0.6" # for databake

Reviewer 2: litemap will be removed with 1.1

Author: (requires 1.1)

Reviewer 2: All our tutorials are being continuously updated, doing it all at release time is a lot of work. Can we somehow release them instead of serving them from head?

Reviewer 2: bump

Author: Done

zerovec = "0.9" # for databake

# for build.rs:
[build-dependencies]
icu = "1.0"
icu_datagen = "1.0"
icu_provider = "1.0"
```

This example has an additional section for auto-generating the data in build.rs. In your build.rs, invoke the ICU4X Datagen API with the set of keys you require. Don't worry; if using databake, you will get a compiler error if you don't specify enough keys.

> Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section and invoke `icu4x-datagen` manually. See [data_management.md](./data_management.md) for more information.

Reviewer: ```suggestion
Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` sectio,  invoke `icu4x-datagen` manually, and potentially check the output in to your version control. See [data_management.md](./data_management.md) for more information.
```

# Setting up your Cargo.toml file for ICU4X

ICU4X makes heavy use of small crates and Cargo features in order to be highly modular. This tutorial is intended to help you set up a Cargo.toml file to download what you need for ICU4X.

## Basic Cargo.toml with testdata

The most basic Cargo.toml to get you off the ground is the following:

```toml
[dependencies]
icu = "1.1"
icu_testdata = "1.1"
```

In your main.rs, you can use all stable ICU4X components in the small number of test locales available in `icu_testdata`.

[« Fully Working Example »](./cargo_tests/testdata)

## Cargo.toml with experimental testdata

If you wish to use an experimental feature, such as `icu_segmenter`, set up your Cargo.toml like this:

```toml
[dependencies]

> icu = { version = "1.1", features = ["experimental"] }

Author: Yes it's fixed in the real Cargo.toml file but I forgot to fix it here. Done

icu_testdata = { version = "1.1", features = ["icu_segmenter"] }
```

> In your main.rs, you can now use experimental features.

Author: done

[« Fully Working Example »](./cargo_tests/experimental)

## Cargo.toml with Buffer Provider

If you wish to generate your own data in blob format and pass it into ICU4X, enable the "serde" Cargo feature as follows:

```toml
[dependencies]
icu = { version = "1.1", features = ["serde"] }
icu_provider_blob = "1.1"
```

> To learn about building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).

Author: done

[« Fully Working Example »](./cargo_tests/buffer)

## Cargo.toml with Baked Provider and build.rs

If you wish to use baked data for ICU4X, add some additional dependencies as follows:

```toml
[dependencies]
icu = "1.1"
icu_provider = "1.1" # for databake
zerovec = "0.9" # for databake

# for build.rs:
[build-dependencies]
icu = "1.1"
icu_datagen = "1.1"

> icu_provider = "1.1"

Author: done

```

This example has an additional section for auto-generating the data in build.rs. In your build.rs, invoke the ICU4X Datagen API with the set of keys you require. Don't worry; if using databake, you will get a compiler error if you don't specify enough keys.

> Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section,  invoke `icu4x-datagen` manually, and potentially check in the output to your version control. See [data_management.md](./data_management.md) for more information.

Reviewer 2: ```suggestion
Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section,  invoke `icu4x-datagen` manually, and check in the output to your version control. See [data_management.md](./data_management.md) for more information.
```

[« Fully Working Example »](./cargo_tests/baked)

## Cargo.toml with Thread Safety (Lazy Static)

If you wish to share ICU4X objects between threads, you must enable the `"sync"` Cargo feature of the `icu_provider` crate, as follows:

```toml
[dependencies]
icu = "1.1"
icu_testdata = "1.1"
icu_provider = { version = "1.1", features = ["sync"] }

> lazy_static = "1.4"

Author: ok I can remove it from here and leave it in the real Cargo.toml

```

> You can now, for example, persist a particular ICU4X object in a [lazy_static](https://docs.rs/lazy_static/latest/lazy_static/).

Reviewer 2: ```suggestion
You can now use most ICU4X types when `Send + Sync` are required, for example to persist them in a [lazy_static](https://docs.rs/lazy_static/latest/lazy_static/).
```

## Cargo.toml with Baked Provider

> It is possible to auto-generate baked data in your build.rs. Use caution with this approach since it will make your build.rs file access the network and therefore be potentially non-deterministic; do not do this if you plan to have other people depending on your crate!

Reviewer: suggestion: we should not say that something is bad without providing an alternative: perhaps include a note that "if you would prefer to not deal with this problem; check in the generated files and regenerate as and when necessary"
